### PR TITLE
8303134: JFR: Missing stack trace during chunk rotation stress

### PR DESCRIPTION
**Pull request by mgronlun**
Greetings,

please help review this fix for some problematic situations in JFR where data can be lost. Most problems originate from writing event data in the wrong epoch due to safepointing. Detailed information about the changes is in the JIRA issue.

Testing: jdk_jfr, stress testing.

Thanks
Markus

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303134](https://bugs.openjdk.org/browse/JDK-8303134): JFR: Missing stack trace during chunk rotation stress (**Bug** - P2)

### Reviewers
 * [Erik Gahlin](https://openjdk.org/census#egahlin) (@egahlin - **Reviewer**) ⚠️ Review applies to [76b27a4d](https://git.openjdk.org/jdk/pull/14761/files/76b27a4de61da033800d58e4facdc2ed66d8757c)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [97854c2e](https://git.openjdk.org/jdk/pull/14761/files/97854c2eca678dc69eb4e0336cb12a61d1d7d9a6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14761/head:pull/14761` \
`$ git checkout pull/14761`

Update a local copy of the PR: \
`$ git checkout pull/14761` \
`$ git pull https://git.openjdk.org/jdk.git pull/14761/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14761`

View PR using the GUI difftool: \
`$ git pr show -t 14761`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14761.diff">https://git.openjdk.org/jdk/pull/14761.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14761#issuecomment-1618993319)

### CODE REVIEW

**Review by egahlin — APPROVED** [review]
Looks good, but the intrinsics need to be reviewed by the compiler team.

File: src/hotspot/share/opto/library_call.cpp [hunk 1]
```diff
@@ -3003,6 +3004,136 @@ bool LibraryCallKit::inline_native_classID() {
...
+
+  // TLS.
+  Node* tls_ptr = _gvn.transform(new ThreadLocalNode());
+  // Jfr java buffer.
+  Node* java_buffer_offset = _gvn.transform(new AddPNode(top(), tls_ptr, _gvn.transform(MakeConX(in_bytes(JAVA_BUFFER_OFFSET_JFR)))));
+  Node* java_buffer = _gvn.transform(new LoadPNode(control(), input_memory_state, java_buffer_offset, TypePtr::BOTTOM, TypeRawPtr::NOTNULL, MemNode::unordered));
+  Node* java_buffer_pos_offset = _gvn.transform(new AddPNode(top(), java_buffer, _gvn.transform(MakeConX(in_bytes(JFR_BUFFER_POS_OFFSET)))));
+
+  // Load the current value of the notified field in the JfrThreadLocal.
+  Node* notified_offset =  basic_plus_adr(top(), tls_ptr, in_bytes(NOTIFY_OFFSET_JFR));
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/library_call.cpp:3024]:
```suggestion
  Node* notified_offset = basic_plus_adr(top(), tls_ptr, in_bytes(NOTIFY_OFFSET_JFR));
```

**Review by TobiHartmann — APPROVED** [review]
The C2 changes look good to me. I guess this needs some testing on platforms not supported by Oracle.

**Comment by RealFYang** [comment]
FYI: Performed `jdk_jfr` test on my linux-riscv64 server, result is clean. Hope that helps.

**Comment by mgronlun** [comment]
> FYI: Performed `jdk_jfr` test on my 64-core linux-riscv64 server, result is clean. Hope that helps.

Thanks Fei, it helps!

**Comment by mgronlun** [comment]
Thank you @egahlin and @TobiHartmann for your reviews!

File: src/hotspot/cpu/s390/stubGenerator_s390.cpp [hunk 2]
```diff
@@ -3133,9 +3140,17 @@ class StubGenerator: public StubCodeGenerator {
     StubRoutines::_cont_returnBarrier = generate_cont_returnBarrier();
     StubRoutines::_cont_returnBarrierExc = generate_cont_returnBarrier_exception();
 
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint_stub = generate_jfr_write_checkpoint();)
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint = StubRoutines::_jfr_write_checkpoint_stub->entry_point();)
+    JFR_ONLY(generate_jfr_stubs();)
+  }
+
+#ifdef INCLUDE_JFR
```
Comment (mgronlun) [inline src/hotspot/cpu/s390/stubGenerator_s390.cpp:3146]:
```suggestion
#ifdef INCLUDE_JFR
```

Comment (mgronlun) [inline src/hotspot/cpu/s390/stubGenerator_s390.cpp:3146]:
```suggestion
#if INCLUDE_JFR
```

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 3]
```diff
@@ -4008,9 +4046,17 @@ class StubGenerator: public StubCodeGenerator {
     StubRoutines::_cont_returnBarrier    = generate_cont_returnBarrier();
     StubRoutines::_cont_returnBarrierExc = generate_cont_returnBarrier_exception();
 
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint_stub = generate_jfr_write_checkpoint();)
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint = StubRoutines::_jfr_write_checkpoint_stub->entry_point();)
+    JFR_ONLY(generate_jfr_stubs();)
+  }
+
+#ifdef INCLUDE_JFR
```
Comment (mgronlun) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:4052]:
```suggestion
#if INCLUDE_JFR
```

File: src/hotspot/cpu/ppc/stubGenerator_ppc.cpp [hunk 4]
```diff
@@ -4728,9 +4763,17 @@ class StubGenerator: public StubCodeGenerator {
     StubRoutines::_cont_returnBarrier = generate_cont_returnBarrier();
     StubRoutines::_cont_returnBarrierExc = generate_cont_returnBarrier_exception();
 
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint_stub = generate_jfr_write_checkpoint();)
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint = StubRoutines::_jfr_write_checkpoint_stub->entry_point();)
+    JFR_ONLY(generate_jfr_stubs();)
+  }
+
+#ifdef INCLUDE_JFR
```
Comment (mgronlun) [inline src/hotspot/cpu/ppc/stubGenerator_ppc.cpp:4769]:
```suggestion
#if INCLUDE_JFR
```

File: src/hotspot/cpu/arm/stubGenerator_arm.cpp [hunk 5]
```diff
@@ -3116,9 +3156,17 @@ class StubGenerator: public StubCodeGenerator {
     StubRoutines::_cont_returnBarrier = generate_cont_returnBarrier();
     StubRoutines::_cont_returnBarrierExc = generate_cont_returnBarrier_exception();
 
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint_stub = generate_jfr_write_checkpoint();)
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint = StubRoutines::_jfr_write_checkpoint_stub->entry_point();)
+    JFR_ONLY(generate_jfr_stubs();)
+  }
+
+#ifdef INCLUDE_JFR
```
Comment (mgronlun) [inline src/hotspot/cpu/arm/stubGenerator_arm.cpp:3162]:
```suggestion
#if INCLUDE_JFR
```

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 6]
```diff
@@ -8261,9 +8299,17 @@ class StubGenerator: public StubCodeGenerator {
     StubRoutines::_cont_returnBarrier = generate_cont_returnBarrier();
     StubRoutines::_cont_returnBarrierExc = generate_cont_returnBarrier_exception();
 
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint_stub = generate_jfr_write_checkpoint();)
-    JFR_ONLY(StubRoutines::_jfr_write_checkpoint = StubRoutines::_jfr_write_checkpoint_stub->entry_point();)
+    JFR_ONLY(generate_jfr_stubs();)
+  }
+
+#ifdef INCLUDE_JFR
```
Comment (mgronlun) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:8305]:
```suggestion
#if INCLUDE_JFR
```

**Pull request closed without merging**